In [1]:
import pandas as pd
import numpy as np

In [2]:
country_names = ['Afghanistan','Aland Islands','Albania','Algeria','American Samoa','Andorra','Angola','Anguilla',
                 'Antarctica','Antigua and Barbuda','Argentina','Armenia','Aruba','Australia','Austria','Azerbaijan',
                 'Bahamas','Bahrain','Bangladesh','Barbados','Belarus','Belgium','Belize','Benin','Bermuda','Bhutan',
                 'Bolivia','Bonaire, Saint Eustatius and Saba ','Bosnia and Herzegovina','Botswana','Bouvet Island',
                 'Brazil','British Indian Ocean Territory','British Virgin Islands','Brunei','Bulgaria','Burkina Faso',
                 'Burundi','Cambodia','Cameroon','Canada','Cape Verde','Cayman Islands','Central African Republic','Chad',
                 'Chile','China','Christmas Island','Cocos Islands','Colombia','Comoros','Cook Islands','Costa Rica',
                 'Croatia','Cuba','Curacao','Cyprus','Czech Republic','Democratic Republic of the Congo','Denmark',
                 'Djibouti','Dominica','Dominican Republic','East Timor','Ecuador','Egypt','El Salvador','Equatorial Guinea',
                 'Eritrea','Estonia','Ethiopia','Falkland Islands','Faroe Islands','Fiji','Finland','France','French Guiana',
                 'French Polynesia','French Southern Territories','Gabon','Gambia','Georgia','Germany','Ghana','Gibraltar',
                 'Greece','Greenland','Grenada','Guadeloupe','Guam','Guatemala','Guernsey','Guinea','Guinea-Bissau','Guyana',
                 'Haiti','Heard Island and McDonald Islands','Honduras','Hong Kong','Hungary','Iceland','India','Indonesia',
                 'Iran','Iraq','Ireland','Isle of Man','Israel','Italy','Ivory Coast','Jamaica','Japan','Jersey','Jordan',
                 'Kazakhstan','Kenya','Kiribati','Kosovo','Kuwait','Kyrgyzstan','Laos','Latvia','Lebanon','Lesotho','Liberia',
                 'Libya','Liechtenstein','Lithuania','Luxembourg','Macao','Macedonia','Madagascar','Malawi','Malaysia',
                 'Maldives','Mali','Malta','Marshall Islands','Martinique','Mauritania','Mauritius','Mayotte','Mexico',
                 'Micronesia','Moldova','Monaco','Mongolia','Montenegro','Montserrat','Morocco','Mozambique','Myanmar',
                 'Namibia','Nauru','Nepal','Netherlands','Netherlands Antilles','New Caledonia','New Zealand','Nicaragua',
                 'Niger','Nigeria','Niue','Norfolk Island','North Korea','Northern Mariana Islands','Norway','Oman','Pakistan',
                 'Palau','Palestinian Territory','Panama','Papua New Guinea','Paraguay','Peru','Philippines','Pitcairn','Poland',
                 'Portugal','Puerto Rico','Qatar','Republic of the Congo','Reunion','Romania','Russia','Rwanda','Saint Barthelemy',
                 'Saint Helena','Saint Kitts and Nevis','Saint Lucia','Saint Martin','Saint Pierre and Miquelon','Saint Vincent and the Grenadines',
                 'Samoa','San Marino','Sao Tome and Principe','Saudi Arabia','Senegal','Serbia','Serbia and Montenegro','Seychelles',
                 'Sierra Leone','Singapore','Sint Maarten','Slovakia','Slovenia','Solomon Islands','Somalia','South Africa','South Georgia and the South Sandwich Islands',
                 'South Korea','South Sudan','Spain','Sri Lanka','Sudan','Suriname','Svalbard and Jan Mayen','Swaziland','Sweden',
                 'Switzerland','Syria','Taiwan','Tajikistan','Tanzania','Thailand','Togo','Tokelau','Tonga','Trinidad and Tobago',
                 'Tunisia','Turkey','Turkmenistan','Turks and Caicos Islands','Tuvalu','U.S. Virgin Islands','Uganda','Ukraine',
                 'United Arab Emirates','United Kingdom','United States','United States Minor Outlying Islands','Uruguay','Uzbekistan',
                 'Vanuatu','Vatican','Venezuela','Vietnam','Wallis and Futuna','Western Sahara','Yemen','Zambia','Zimbabwe','World', 'Rest of World']

### Fertility vs mortality

In [3]:
fert_vs_mort = pd.read_csv("../data/original_data/fertility-vs-child-mortality.csv")

In [4]:
fert_vs_mort.head()

Entity Code  Year  Child mortality rate (%)  \
0  Afghanistan  AFG  1800                       NaN   
1  Afghanistan  AFG  1801                       NaN   
2  Afghanistan  AFG  1802                       NaN   
3  Afghanistan  AFG  1803                       NaN   
4  Afghanistan  AFG  1804                       NaN   

   Fertility Rate (Children per woman) (live births per woman)  \
0                                                NaN             
1                                                NaN             
2                                                NaN             
3                                                NaN             
4                                                NaN             

   Total population (Gapminder)  
0                     3280000.0  
1                     3280000.0  
2                     3280000.0  
3                     3280000.0  
4                     3280000.0

In [5]:
fert_vs_mort.drop(columns = "Total population (Gapminder)", inplace = True)

In [6]:
fert_vs_mort.rename(columns = {"Entity" : "Country", "Fertility Rate (Children per woman) (live births per woman)" : "Fertility Rate"}, inplace = True)

In [7]:
fert_vs_mort.dtypes

Country                      object
Code                         object
Year                         object
Child mortality rate (%)    float64
Fertility Rate              float64
dtype: object

In [8]:
fert_vs_mort.isna().sum()

Country                         0
Code                         3112
Year                            0
Child mortality rate (%)    33419
Fertility Rate              33419
dtype: int64

In [9]:
fert_vs_mort = fert_vs_mort.loc[fert_vs_mort["Country"].isin(country_names)]

In [10]:
fert_vs_mort.isna().sum()

Country                         0
Code                            0
Year                            0
Child mortality rate (%)    31669
Fertility Rate              31669
dtype: int64

In [11]:
fert_vs_mort.dropna(inplace = True)

In [12]:
fert_vs_mort["Year"] = pd.to_datetime(fert_vs_mort["Year"], format = "%Y")

In [13]:
fert_vs_mort[["Child mortality rate (%)", "Fertility Rate"]] = fert_vs_mort[["Child mortality rate (%)", "Fertility Rate"]].round(decimals = 2)

In [14]:
fert_vs_mort.head()

Country Code       Year  Child mortality rate (%)  Fertility Rate
150  Afghanistan  AFG 1950-01-01                     41.99            7.45
151  Afghanistan  AFG 1951-01-01                     41.63            7.45
152  Afghanistan  AFG 1952-01-01                     40.93            7.45
153  Afghanistan  AFG 1953-01-01                     40.24            7.45
154  Afghanistan  AFG 1954-01-01                     39.58            7.45

In [15]:
fert_vs_mort.to_csv("../data/fertility-vs-mortality.csv", index = False)

----

### Life expectancy

In [16]:
life_exp = pd.read_csv("../data/original_data/life-expectancy-of-women-vs-life-expectancy-of-women.csv")

In [17]:
life_exp.head()

Entity Code  Year  Life expectancy of men (years)  \
0  Afghanistan  AFG  1800                             NaN   
1  Afghanistan  AFG  1801                             NaN   
2  Afghanistan  AFG  1802                             NaN   
3  Afghanistan  AFG  1803                             NaN   
4  Afghanistan  AFG  1804                             NaN   

   Life expectancy of women (years)  Total population (Gapminder)  
0                               NaN                     3280000.0  
1                               NaN                     3280000.0  
2                               NaN                     3280000.0  
3                               NaN                     3280000.0  
4                               NaN                     3280000.0

In [18]:
life_exp.isna().sum()

Entity                                  0
Code                                 3112
Year                                    0
Life expectancy of men (years)      33419
Life expectancy of women (years)    33419
Total population (Gapminder)         2442
dtype: int64

In [19]:
life_exp.shape

(49325, 6)

Life expenctancy has the same problem with missing value as fertility vs mortality. 
Values missing seems to be from 1800s and has no big impact in this project. 

In [20]:
life_exp.rename(columns = {"Entity" : "Country", "Life expectancy of men (years)" : "Life expectancy men", "Life expectancy of women (years)" : "Life expectancy women"}, inplace = True)

In [21]:
life_exp.drop(columns = "Total population (Gapminder)", inplace = True)

In [22]:
life_exp = life_exp[life_exp["Country"].isin(country_names)]

In [23]:
life_exp.isna().sum()

Country                      0
Code                         0
Year                         0
Life expectancy men      31669
Life expectancy women    31669
dtype: int64

In [24]:
life_exp.head()

Country Code  Year  Life expectancy men  Life expectancy women
0  Afghanistan  AFG  1800                  NaN                    NaN
1  Afghanistan  AFG  1801                  NaN                    NaN
2  Afghanistan  AFG  1802                  NaN                    NaN
3  Afghanistan  AFG  1803                  NaN                    NaN
4  Afghanistan  AFG  1804                  NaN                    NaN

In [25]:
life_exp.dropna(inplace = True)

In [26]:
life_exp

Country Code  Year  Life expectancy men  Life expectancy women
150    Afghanistan  AFG  1950               26.867                 28.360
151    Afghanistan  AFG  1951               27.138                 28.632
152    Afghanistan  AFG  1952               27.675                 29.172
153    Afghanistan  AFG  1953               28.201                 29.703
154    Afghanistan  AFG  1954               28.716                 30.223
...            ...  ...   ...                  ...                    ...
49316     Zimbabwe  ZWE  2011               53.425                 56.033
49317     Zimbabwe  ZWE  2012               55.036                 57.848
49318     Zimbabwe  ZWE  2013               56.461                 59.494
49319     Zimbabwe  ZWE  2014               57.657                 60.911
49320     Zimbabwe  ZWE  2015               58.593                 62.050

[12672 rows x 5 columns]

In [27]:
life_exp["Year"] = pd.to_datetime(life_exp["Year"], format = "%Y")

In [28]:
life_exp[["Life expectancy men", "Life expectancy women"]] = life_exp[["Life expectancy men", "Life expectancy women"]].round(decimals = 2)

In [29]:
life_exp.dtypes

Country                          object
Code                             object
Year                     datetime64[ns]
Life expectancy men             float64
Life expectancy women           float64
dtype: object

In [30]:
life_exp.head()

Country Code       Year  Life expectancy men  Life expectancy women
150  Afghanistan  AFG 1950-01-01                26.87                  28.36
151  Afghanistan  AFG 1951-01-01                27.14                  28.63
152  Afghanistan  AFG 1952-01-01                27.68                  29.17
153  Afghanistan  AFG 1953-01-01                28.20                  29.70
154  Afghanistan  AFG 1954-01-01                28.72                  30.22

In [31]:
life_exp.to_csv("../data/life_expectancy.csv", index = False)

----

### GDP

In [32]:
gdp = pd.read_csv("../data/original_data/gdp.csv")

In [33]:
gdp.head()

Code      Country                       Subject Descriptor  \
0  AFG  Afghanistan  Gross domestic product, constant prices   
1  AFG  Afghanistan  Gross domestic product, constant prices   
2  AFG  Afghanistan   Gross domestic product, current prices   
3  AFG  Afghanistan   Gross domestic product, current prices   
4  AFG  Afghanistan   Gross domestic product, current prices   

                                       Subject Notes  \
0  Expressed in billions of national currency uni...   
1  Annual percentages of constant price GDP are y...   
2  Expressed in billions of national currency uni...   
3  Values are based upon GDP in national currency...   
4  These data form the basis for the country weig...   

                                            Units     Scale  \
0                               National currency  Billions   
1                                  Percent change       NaN   
2                               National currency  Billions   
3                                    U.S. dollars  Billions   
4  Purchasing power parity; international dollars  Billions   

                       Country/Series-specific Notes 1980 1981 1982  ...  \
0  Source: National Statistics Office Latest actu...  NaN  NaN  NaN  ...   
1  See notes for:  Gross domestic product, consta...  NaN  NaN  NaN  ...   
2  Source: National Statistics Office Latest actu...  NaN  NaN  NaN  ...   
3  See notes for:  Gross domestic product, curren...  NaN  NaN  NaN  ...   
4  See notes for:  Gross domestic product, curren...  NaN  NaN  NaN  ...   

       2015      2016      2017      2018      2019      2020      2021  \
0   482.629   493.073   506.215   519.883   535.479   554.101   576.265   
1     0.988     2.164     2.665       2.7         3     3.478         4   
2  1,226.57  1,318.48  1,377.54  1,421.41  1,502.47  1,624.89  1,770.34   
3    20.057    19.428    20.235     19.63    18.734    18.861    19.998   
4    64.231    66.301     69.35    72.958    76.486    80.717    85.678   

       2022      2023      2024  
0   602.197   632.307   667.084  
1       4.5         5       5.5  
2  1,942.73  2,142.06  2,373.04  
3     21.54    23.237     25.19  
4    91.329    97.814   105.271  

[5 rows x 52 columns]

In [34]:
gdp = gdp[gdp["Country"].isin(country_names)]

In [35]:
gdp["Subject Descriptor"].value_counts()

Gross domestic product per capita, current prices                                     519
Gross domestic product, current prices                                                519
General government net debt                                                           346
General government gross debt                                                         346
General government primary net lending/borrowing                                      346
Gross domestic product per capita, constant prices                                    346
General government structural balance                                                 346
Gross domestic product, constant prices                                               346
General government net lending/borrowing                                              346
General government total expenditure                                                  346
General government revenue                                                            346
Inflation,

In [36]:
gdp = gdp.loc[(gdp["Subject Descriptor"] == "Gross domestic product per capita, current prices") | (gdp["Subject Descriptor"] == "Inflation, average consumer prices") | (gdp["Subject Descriptor"] == "Unemployment rate") ]

In [37]:
gdp.loc[gdp["Country"] == "India"]

Code Country                                 Subject Descriptor  \
3428  IND   India  Gross domestic product per capita, current prices   
3429  IND   India  Gross domestic product per capita, current prices   
3430  IND   India  Gross domestic product per capita, current prices   
3436  IND   India                 Inflation, average consumer prices   
3437  IND   India                 Inflation, average consumer prices   
3445  IND   India                                  Unemployment rate   

                                          Subject Notes  \
3428  GDP is expressed in current national currency ...   
3429  GDP is expressed in current U.S. dollars per p...   
3430  Expressed in GDP in PPP dollars per person. Da...   
3436  Expressed in averages for the year, not end-of...   
3437  Annual percentages of average consumer prices ...   
3445  Unemployment rate can be defined by either the...   

                                               Units  Scale  \
3428                               National currency  Units   
3429                                    U.S. dollars  Units   
3430  Purchasing power parity; international dollars  Units   
3436                                           Index    NaN   
3437                                  Percent change    NaN   
3445                    Percent of total labor force    NaN   

                          Country/Series-specific Notes      1980      1981  \
3428  See notes for:  Gross domestic product, curren...  2,182.96  2,511.51   
3429  See notes for:  Gross domestic product, curren...    276.35   280.764   
3430  See notes for:  Gross domestic product, curren...     558.9   635.088   
3436  Source: National Statistics Office. The origin...      8.96    10.094   
3437  See notes for:  Inflation, average consumer pr...    11.346    12.654   
3445                                                NaN       NaN       NaN   

          1982  ...        2015        2016        2017        2018  \
3428  2,748.34  ...  107,348.01  118,190.78  129,812.92  142,481.36   
3429   284.468  ...    1,639.69    1,761.63    2,014.01    2,037.69   
3430   682.651  ...    6,259.80    6,752.48    7,277.04    7,858.69   
3436    10.875  ...     132.323     138.278     143.259     148.171   
3437     7.736  ...         4.9         4.5       3.602       3.428   
3445       NaN  ...         NaN         NaN         NaN         NaN   

            2019        2020        2021        2022        2023        2024  
3428  154,600.53  169,780.41  187,034.75  206,161.73  227,236.49  250,213.09  
3429    2,171.64    2,338.12    2,529.34    2,737.15    2,965.51    3,209.67  
3430    8,378.38    9,026.87    9,770.27   10,568.27   11,431.35   12,353.81  
3436     153.262     159.525     166.022     172.737     179.618     186.756  
3437       3.436       4.086       4.073       4.045       3.983       3.974  
3445         NaN         NaN         NaN         NaN         NaN         NaN  

[6 rows x 52 columns]

In [38]:
gdp["Subject Descriptor"].value_counts()

Gross domestic product per capita, current prices    519
Inflation, average consumer prices                   346
Unemployment rate                                    173
Name: Subject Descriptor, dtype: int64

In [39]:
gdp["Units"].value_counts()

Percent of total labor force                      173
U.S. dollars                                      173
Index                                             173
Percent change                                    173
Purchasing power parity; international dollars    173
National currency                                 173
Name: Units, dtype: int64

In [40]:
gdp.drop(columns = ["Subject Notes", "Scale", "Country/Series-specific Notes"], inplace = True)

In [41]:
gdp.isna().sum()

Code                    0
Country                 0
Subject Descriptor      0
Units                   0
1980                  370
1981                  359
1982                  357
1983                  353
1984                  351
1985                  347
1986                  343
1987                  341
1988                  339
1989                  337
1990                  303
1991                  291
1992                  231
1993                  211
1994                  202
1995                  178
1996                  167
1997                  155
1998                  143
1999                  137
2000                  117
2001                  111
2002                  103
2003                   97
2004                   90
2005                   88
2006                   88
2007                   83
2008                   83
2009                   83
2010                   82
2011                   84
2012                   83
2013                   83
2014        

In [42]:
gdp_new = gdp.loc[(gdp["Units"] == "U.S. dollars") | (gdp["Units"] == "Percent change") | (gdp["Units"] == "Percent of total labor force")]

In [43]:
gdp_new.loc[gdp_new["Country"] == "China"]

Code Country                                 Subject Descriptor  \
1584  CHN   China  Gross domestic product per capita, current prices   
1592  CHN   China                 Inflation, average consumer prices   
1600  CHN   China                                  Unemployment rate   

                             Units     1980     1981     1982     1983  \
1584                  U.S. dollars  309.354  290.617  282.054  298.692   
1592                Percent change      NaN      2.5        2        2   
1600  Percent of total labor force      4.9      3.8      3.2      2.3   

         1984     1985  ...      2015      2016      2017      2018  \
1584  303.408  295.336  ...  8,166.76  8,115.83  8,677.40  9,580.24   
1592      2.7      9.3  ...     1.441     2.003     1.558       2.1   
1600      1.9      1.8  ...      4.05      4.02       3.9       3.8   

           2019       2020       2021       2022       2023       2024  
1584  10,098.87  10,872.50  11,773.10  12,718.79  13,723.67  14,811.79  
1592      2.305      2.432        2.8        2.9          3          3  
1600        3.8        3.8        3.8        3.8        3.8        3.8  

[3 rows x 49 columns]

In [44]:
gdp_new.columns

Index(['Code', 'Country', 'Subject Descriptor', 'Units', '1980', '1981',
       '1982', '1983', '1984', '1985', '1986', '1987', '1988', '1989', '1990',
       '1991', '1992', '1993', '1994', '1995', '1996', '1997', '1998', '1999',
       '2000', '2001', '2002', '2003', '2004', '2005', '2006', '2007', '2008',
       '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016', '2017',
       '2018', '2019', '2020', '2021', '2022', '2023', '2024'],
      dtype='object')

In [45]:
gdp_new.head()

Code      Country                                 Subject Descriptor  \
9   AFG  Afghanistan  Gross domestic product per capita, current prices   
17  AFG  Afghanistan                 Inflation, average consumer prices   
25  AFG  Afghanistan                                  Unemployment rate   
54  ALB      Albania  Gross domestic product per capita, current prices   
62  ALB      Albania                 Inflation, average consumer prices   

                           Units     1980     1981     1982     1983     1984  \
9                   U.S. dollars      NaN      NaN      NaN      NaN      NaN   
17                Percent change      NaN      NaN      NaN      NaN      NaN   
25  Percent of total labor force      NaN      NaN      NaN      NaN      NaN   
54                  U.S. dollars  728.359  817.734  824.542  815.529  788.332   
62                Percent change      NaN      NaN      NaN      NaN      NaN   

       1985  ...      2015      2016      2017      2018      2019      2020  \
9       NaN  ...   594.513   560.601   569.531   544.983   513.108   509.759   
17      NaN  ...    -0.662     4.384     4.976     0.626     2.587     4.502   
25      NaN  ...       NaN       NaN       NaN       NaN       NaN       NaN   
54  788.801  ...  3,953.61  4,124.41  4,540.46  5,239.21  5,372.74  5,847.06   
62      NaN  ...     1.868     1.282     1.987     2.029     1.799         2   

        2021      2022      2023      2024  
9    533.089   566.416   602.884    644.95  
17     4.748         5         5         5  
25       NaN       NaN       NaN       NaN  
54  6,333.43  6,876.57  7,410.75  7,993.47  
62       2.4       2.8         3         3  

[5 rows x 49 columns]

#### Goal:

|Country|Code|Year|GDP per capita|Inflation percent| Unemployment rate|
|-------|----|----|--------------|-----------------|------------------|
|Afghan |AFG |2015| 594.513      |  -0.662         |  NaN             |
|Afghan |AFG |2016| 560.601      |   4.384         |  NaN             |
|...    |... |


In [46]:
pd.melt(gdp_new, id_vars = ["Code", "Country"], value_vars = ['1980', '1981',
       '1982', '1983', '1984', '1985', '1986', '1987', '1988', '1989', '1990',
       '1991', '1992', '1993', '1994', '1995', '1996', '1997', '1998', '1999',
       '2000', '2001', '2002', '2003', '2004', '2005', '2006', '2007', '2008',
       '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016', '2017',
       '2018', '2019', '2020', '2021', '2022', '2023', '2024'], var_name = "Year")

Code      Country  Year    value
0      AFG  Afghanistan  1980      NaN
1      AFG  Afghanistan  1980      NaN
2      AFG  Afghanistan  1980      NaN
3      ALB      Albania  1980  728.359
4      ALB      Albania  1980      NaN
...    ...          ...   ...      ...
23350  ZMB       Zambia  2024        8
23351  ZMB       Zambia  2024      NaN
23352  ZWE     Zimbabwe  2024  971.081
23353  ZWE     Zimbabwe  2024        3
23354  ZWE     Zimbabwe  2024      NaN

[23355 rows x 4 columns]

Using melt works, but I'll need to create separate dfs for each `Subject Descriptor`

In [47]:
gdp_new.drop(columns = "Units", inplace = True)

C:\Anaconda\envs\new_env\lib\site-packages\pandas\core\frame.py:3990: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [48]:
gdp_gdp = gdp_new.loc[gdp_new["Subject Descriptor"] == "Gross domestic product per capita, current prices"]

In [49]:
gdp_infla = gdp_new.loc[gdp_new["Subject Descriptor"] == "Inflation, average consumer prices"]

In [50]:
gdp_unem = gdp_new.loc[gdp_new["Subject Descriptor"] == "Unemployment rate"]

In [51]:
gdp_unem.loc[gdp_unem["Country"] == "India"]

Code Country Subject Descriptor 1980 1981 1982 1983 1984 1985 1986  ...  \
3445  IND   India  Unemployment rate  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...   

     2015 2016 2017 2018 2019 2020 2021 2022 2023 2024  
3445  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[1 rows x 48 columns]

No information for unemployment rate for India.

In [52]:
gdp_pivot = pd.melt(gdp_gdp, 
                    id_vars = ["Code", "Country"], 
                    value_vars = ['1980', '1981',
                                  '1982', '1983', '1984', '1985', '1986', '1987', '1988', '1989', '1990',
                                  '1991', '1992', '1993', '1994', '1995', '1996', '1997', '1998', '1999',
                                  '2000', '2001', '2002', '2003', '2004', '2005', '2006', '2007', '2008',
                                  '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016', '2017',
                                  '2018', '2019', '2020', '2021', '2022', '2023', '2024'],
                    var_name = "Year", 
                    value_name = "GDP per capita (USD)")

In [53]:
gdp_pivot.shape

(7785, 4)

In [54]:
infla_pivot = pd.melt(gdp_infla, 
                      id_vars = ["Code", "Country"], 
                      value_vars = ['1980', '1981',
                                    '1982', '1983', '1984', '1985', '1986', '1987', '1988', '1989', '1990',
                                    '1991', '1992', '1993', '1994', '1995', '1996', '1997', '1998', '1999',
                                    '2000', '2001', '2002', '2003', '2004', '2005', '2006', '2007', '2008',
                                    '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016', '2017',
                                    '2018', '2019', '2020', '2021', '2022', '2023', '2024'],
                      var_name = "Year", 
                      value_name = "Inflation, avg cons. prices (% change)")

In [55]:
infla_pivot.shape

(7785, 4)

In [56]:
unemp_pivot = pd.melt(gdp_unem, 
                      id_vars = ["Code", "Country"], 
                      value_vars = ['1980', '1981',
                                    '1982', '1983', '1984', '1985', '1986', '1987', '1988', '1989', '1990',
                                    '1991', '1992', '1993', '1994', '1995', '1996', '1997', '1998', '1999',
                                    '2000', '2001', '2002', '2003', '2004', '2005', '2006', '2007', '2008',
                                    '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016', '2017',
                                    '2018', '2019', '2020', '2021', '2022', '2023', '2024'],
                      var_name = "Year", 
                      value_name = "Unemployment rate")

In [57]:
unemp_pivot.shape

(7785, 4)

In [58]:
gdp_merged = gdp_pivot.merge(infla_pivot, on = ["Code", "Country", "Year"], how = "left")

In [59]:
gdp_merged = gdp_merged.merge(unemp_pivot, on = ["Code", "Country", "Year"], how = "left")

Now that the dataframe looks like I want, is time to finish the cleaning. 

In [60]:
gdp_merged.dtypes

Code                                      object
Country                                   object
Year                                      object
GDP per capita (USD)                      object
Inflation, avg cons. prices (% change)    object
Unemployment rate                         object
dtype: object

In [61]:
gdp_merged["Year"] = pd.to_datetime(gdp_merged["Year"], format = "%Y")

In [62]:
gdp_merged.head()

Code              Country       Year GDP per capita (USD)  \
0  AFG          Afghanistan 1980-01-01                  NaN   
1  ALB              Albania 1980-01-01              728.359   
2  DZA              Algeria 1980-01-01             2,268.61   
3  AGO               Angola 1980-01-01              745.119   
4  ATG  Antigua and Barbuda 1980-01-01             1,920.27   

  Inflation, avg cons. prices (% change) Unemployment rate  
0                                    NaN               NaN  
1                                    NaN             5.028  
2                                  9.668            15.789  
3                                 46.708               NaN  
4                                 18.999               NaN

In [63]:
type(gdp_merged["GDP per capita (USD)"])

pandas.core.series.Series

In [64]:
gdp_merged["GDP per capita (USD)"] = gdp_merged["GDP per capita (USD)"].str.replace(",", "")

In [65]:
gdp_merged["GDP per capita (USD)"] = pd.to_numeric(gdp_merged["GDP per capita (USD)"])

In [66]:
gdp_merged["Inflation, avg cons. prices (% change)"] = gdp_merged["Inflation, avg cons. prices (% change)"].str.replace("--", "").str.replace(",", "")

In [67]:
gdp_merged["Inflation, avg cons. prices (% change)"] = pd.to_numeric(gdp_merged["Inflation, avg cons. prices (% change)"])

In [68]:
gdp_merged["Unemployment rate"] = pd.to_numeric(gdp_merged["Unemployment rate"])

In [69]:
gdp_merged.head()

Code              Country       Year  GDP per capita (USD)  \
0  AFG          Afghanistan 1980-01-01                   NaN   
1  ALB              Albania 1980-01-01               728.359   
2  DZA              Algeria 1980-01-01              2268.610   
3  AGO               Angola 1980-01-01               745.119   
4  ATG  Antigua and Barbuda 1980-01-01              1920.270   

   Inflation, avg cons. prices (% change)  Unemployment rate  
0                                     NaN                NaN  
1                                     NaN              5.028  
2                                   9.668             15.789  
3                                  46.708                NaN  
4                                  18.999                NaN

In [70]:
gdp_merged.dtypes

Code                                              object
Country                                           object
Year                                      datetime64[ns]
GDP per capita (USD)                             float64
Inflation, avg cons. prices (% change)           float64
Unemployment rate                                float64
dtype: object

In [71]:
gdp_merged["Unemployment rate"] = gdp_merged["Unemployment rate"].round(decimals = 2)

In [72]:
gdp_merged["Inflation, avg cons. prices (% change)"] = gdp_merged["Inflation, avg cons. prices (% change)"].round(decimals = 2)

In [73]:
gdp_merged["GDP per capita (USD)"] = gdp_merged["GDP per capita (USD)"].round(decimals = 2)

In [74]:
gdp_merged.head()

Code              Country       Year  GDP per capita (USD)  \
0  AFG          Afghanistan 1980-01-01                   NaN   
1  ALB              Albania 1980-01-01                728.36   
2  DZA              Algeria 1980-01-01               2268.61   
3  AGO               Angola 1980-01-01                745.12   
4  ATG  Antigua and Barbuda 1980-01-01               1920.27   

   Inflation, avg cons. prices (% change)  Unemployment rate  
0                                     NaN                NaN  
1                                     NaN               5.03  
2                                    9.67              15.79  
3                                   46.71                NaN  
4                                   19.00                NaN

In [75]:
gdp_merged.to_csv("../data/economic-information.csv", index = False)